### ISSUE TITLE
Add `pure` setting to `python_installation` object

### ISSUE THREAD

**Issue by rgommers** [mesonbuild/meson#10523]
tl;dr of this proposal is: make it possible for Python projects to set `pure` in a single place, rather than have to repeat `pure: false` over and over.

This follows up on https://github.com/mesonbuild/meson/pull/10514#issuecomment-1161405026 (Cc @eli-schwartz)

### What is the problem?

For SciPy we currently have 99 lines of `pure: false` in `meson.build` files. For each subdir, we typically have one instance of:
```meson

python_sources = [
  '__init__.py',
  'xxx.py',  # list of file names
]

py3.install_sources(
  python_sources,
  pure: false,
  subdir: 'scipy/some_subdir'
)
```

That `pure: false` boilerplate not only gets repetitive, there are two other issues with it:
1. If you forget it, things will still work in CI, and then on some random Linux distro that happens to still define `purelib` and `platlib` to different directories, the install will be broken at runtime (because Python packages should be in a single source tree, not split - I have seen this issue once already).
2. It makes the average maintainer/contributor, who will not know what `purelib`/`platlib` is about, wonder whether this is relevant - waste of time.

### What does `pure` mean again?

The best description is probably https://peps.python.org/pep-0491/#what-s-the-deal-with-purelib-vs-platlib.

 The last sentence says: _"In practice a wheel should have only one of “purelib” or “platlib” depending on whether it is pure Python or not and those files should be at the root with the appropriate setting given for “Root-is-purelib”."_ This makes sense; and explains why things break if you forget `pure: false` once in a Python package that contains compiled code - the structure of the Python package gets lost if files get scattered over both `purelib` _and_ `platlib`.

Let me also emphasize that the whole purelib/platlib distinction is Linux-only, for a minority of Linux distros, and a leftover from a time where people jumped through hoops to save a little bit of disk space. It makes very little sense as a design in today's context, and should be hidden as much as possible (unfortunately it cannot be removed from Python packaging completely).

### How can we improve things here in Meson?

Because all files for a complete Python package need to have the same `pure` setting, it makes sense for that setting to live on the `python_installation` object, rather than having to specify it every time you use `python_installation.install_sources`. I'm not sure what the best way of adding it there; if the typical code in the top-level `meson.build` looks like:
```meson
py_mod = import('python')
py3 = py_mod.find_installation()
py3_dep = py3.dependency()
```
then it could be a keyword to `find_installation()`, or a separate method like `py3 = py3.set_purelib('false')`.

It should be possible to add something like this in a backwards-compatible way; if it's not used then `install_sources` continues to default to `true`, and if it is given then that overwrites the default.

**Comment by rgommers** [comment]
A somewhat similar situation happens with the `install: true` argument to `extension_module()`. That is not nearly as much of a problem, because if you forget it things break immediately rather than at runtime on a few Linux distros only - but it's also a lot of boilerplate. For `py3.extension_module`, in any Python package you are likely to want to install every single extension, but you now have to type `install: true` on every line. Adding it to the `python_installation` object once as a default could be considered, if my proposal for `pure` above is acceptable.

**Comment by eli-schwartz** [comment]
I'm not 100% sure what I think about this, but if we do implement this I think we'd probably want to do it as a kwarg on find_installation(), rather than a method that mutates an existing found installation.

Aside: another possible solution is implementing some `functools.partial`-like method of preparing a function with set arguments, a topic that has come up a few times as well in relation to generator() being an inferior solution to custom_target in a for loop in many cases.
